flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/KvStateService.java

```diff
@@ -0,0 +1,192 @@
+				}
+			}
+		}
+	}
+
+	public void shutdown() {
+		synchronized (lock) {
+			if (isShutdown) {
+				return;
+			}
+
+			LOG.info("Shutting down the KvState service and its components.");
+
+			if (kvStateClientProxy != null) {
+				try {
+					LOG.debug("Shutting down Queryable State Client Proxy.");
+					kvStateClientProxy.shutdown();
+				} catch (Throwable t) {
+					LOG.warn("Cannot shut down Queryable State Client Proxy.", t);
+				}
+			}
+
+			if (kvStateServer != null) {
+				try {
+					LOG.debug("Shutting down Queryable State Data Server.");
+					kvStateServer.shutdown();
+				} catch (Throwable t) {
+					LOG.warn("Cannot shut down Queryable State Data Server.", t);
+				}
+			}
+
+			isShutdown = true;
+		}
+	}
+
+	// --------------------------------------------------------------------------------------------
+	//  Static factory methods for kvState service
+	// --------------------------------------------------------------------------------------------
+
+	/**
+	 * Creates and returns the KvState service.
+	 *
+	 * @param taskManagerServicesConfiguration task manager configuration
+	 * @return service for kvState related components
+	 */
+	public static KvStateService fromConfiguration(TaskManagerServicesConfiguration taskManagerServicesConfiguration) {
+		KvStateRegistry kvStateRegistry = new KvStateRegistry();
+
+		QueryableStateConfiguration qsConfig = taskManagerServicesConfiguration.getQueryableStateConfig();
+
+		KvStateClientProxy kvClientProxy = null;
+		KvStateServer kvStateServer = null;
+
+		if (qsConfig != null) {
+			int numProxyServerNetworkThreads = qsConfig.numProxyServerThreads() == 0 ?
+				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numProxyServerThreads();
+			int numProxyServerQueryThreads = qsConfig.numProxyQueryThreads() == 0 ?
+				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numProxyQueryThreads();
+			kvClientProxy = QueryableStateUtils.createKvStateClientProxy(
+				taskManagerServicesConfiguration.getTaskManagerAddress(),
+				qsConfig.getProxyPortRange(),
+				numProxyServerNetworkThreads,
+				numProxyServerQueryThreads,
+				new DisabledKvStateRequestStats());
+
+			int numStateServerNetworkThreads = qsConfig.numStateServerThreads() == 0 ?
+				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numStateServerThreads();
+			int numStateServerQueryThreads = qsConfig.numStateQueryThreads() == 0 ?
+				taskManagerServicesConfiguration.getNumberOfSlots() : qsConfig.numStateQueryThreads();
+			kvStateServer = QueryableStateUtils.createKvStateServer(
+				taskManagerServicesConfiguration.getTaskManagerAddress(),
+				qsConfig.getStateServerPortRange(),
+				numStateServerNetworkThreads,
+				numStateServerQueryThreads,
+				kvStateRegistry,
+				new DisabledKvStateRequestStats());
+		}
+
+		return new KvStateService(kvStateRegistry, kvStateServer, kvClientProxy);
+	}
+
+	public static KvStateService build() {
```

Comment:
If this method is just for tests, I would suggest to move to the test code. Maybe, `TaskManagerServicesBuilder` could have a static method `createStubKvStateService`. It is a bit confusing to see this method here in production code.

Reply:
I am a bit wondering it would destroy the `builder` structure for `TaskManagerServiceBuilder`, and considering it is also easy for constructing `KvStateService` directly in tests, so I remove this `build` method. :)